### ISSUE TITLE
modify method exceptions are not catched

### ISSUE THREAD

**Issue by ozum** [tgriesser/knex#1529]
Hi,

Exceptions in `modify` method are not catched by `.catch`

```js
knex.select()
    .from('Account')
    .modify(function(qb) {
        throw new Error('ERROR HERE');
    })
    .then(r => {})
    .catch(err => console.log(err));
```

Regards,

**Comment by elhigu** [comment]
@ozum  that sounds like bug... It really hasn't been documented which errors should be passed to promise chain and which are just thrown out from query builder, but I agree that this is pretty inconsistent, since the error in modify can't be thrown before query is being compiled / executed.

**Comment by ozum** [comment]
This results error silently ignored and request get hung in Hapi.js handler. Do you have any workaround suggestion?

```js
function handlerFunc(request, reply) {
    let query = knex.select().from('some_table');

    return query
        .where('Address.organizationId', 1)
        .modify(rqlToKnex, rql)
        .then(r => {
            return reply({
                result: r
            });
        })
        .catch(err => reply(err));
}
```

**Comment by elhigu** [comment]
Maybe adding try catch to get exception from modify... pretty ugly.

```js
function handlerFunc(request, reply) {
    let query = knex.select().from('some_table');

    try {
      return query
          .where('Address.organizationId', 1)
          .modify(rqlToKnex, rql)
          .then(r => {
              return reply({
                  result: r
              });
          })
          .catch(err => reply(err));
    } catch(err) {
      return Promise.resolve(reply(err));
    }
}

server.route({
    method: 'GET',
    path: '/accounts',
    handler: handlerFunc
});
```

Also throwing exception inside `then` callback should catch the error and pass it to promise chain... so this looks nicer and should work too:

```js
function handlerFunc(request, reply) {
  let query = knex.select().from('some_table');

  return Promise.resolve()
    .then(() => {
      return query
        .where('Address.organizationId', 1)
        .modify(rqlToKnex, rql);
    })
    .then(r => reply({ result: r }))
    .catch(err => reply(err));
}

server.route({
  method: 'GET',
  path: '/accounts',
  handler: handlerFunc
});
```

**Comment by ozum** [comment]
You are right, I also couldn't find a beautiful workaround. My code has lots of `modify` calls, and I hope a fix for `modify` will available soon.

Thanks.

**Comment by elhigu** [comment]
To me the second workaround was actually pretty nice :) One more workaround would be to not throw exceptions from modify...

**Comment by johanneslumpe** [comment]
@ozum Another "workaround" might be to convert the code to use `async/await` and then register a custom `async` handler in hapi, which can catch errors and handle them in a single spot.

**Comment by johanneslumpe** [comment]
@elhigu Your second version could be even nicer by just using `Promise.try` (when using bluebird):

```js
return Promise
  .try(() => (
    knex
      .select()
      .from('some_table')
      .where('Address.organizationId', 1)
      .modify(rqlToKnex, rql)
  ))
  .then(r => reply({ result: r }))
  .catch(err => reply(err));  
```

**Comment by ozum** [comment]
@elhigu, I mean, I could not find a workaround without adding further nested code, otherwise second workaround is totally acceptable, and I already applied second one in a couple of places. However for another larger project I have, I wish I had small number of modify methods.

@johanneslumpe, also nice suggestions.

Thanks both of you guys for your helps and suggestions.

**Comment by mercmobily** [comment]
You're absolutely right that this inconsistency still exists, and I can confirm it remains an issue in the current codebase. This is indeed a valid bug that should be addressed.

The problem is that the `modify` method executes its callback function synchronously at call time:

```javascript
modify(callback) {
  callback.apply(this, [this].concat(tail(arguments)));
  return this;
}
```

This means any exceptions thrown inside the callback are thrown immediately, before the promise chain is established. In contrast, methods like `where` with function arguments use `whereWrapped`, which stores the callback to be executed later during query compilation, allowing errors to be properly caught by the promise chain.

This behavior is inconsistent and confusing for developers. The issue has valid workarounds (as discussed in the comments), but the inconsistency between `modify` and other query builder methods creates unexpected behavior.

The cleanest fix would be to modify the `modify` method to store callbacks for later execution during query compilation, similar to how `whereWrapped` works. Alternatively, we could wrap the callback execution in a try-catch and store any errors to be thrown during query execution.

For users hitting this issue right now, the recommended workarounds are:

```javascript
// Option 1: Wrap in Promise.resolve() 
return Promise.resolve()
  .then(() => {
    return query
      .where('someColumn', 1)
      .modify(rqlToKnex, rql);
  })
  .then(r => reply({ result: r }))
  .catch(err => reply(err));

// Option 2: Use async/await with try-catch
try {
  const result = await query
    .where('someColumn', 1)
    .modify(rqlToKnex, rql);
  return reply({ result });
} catch (err) {
  return reply(err);
}
```

This is a breaking behaviour inconsistency that affects error handling patterns. While it's been open since 2016, it represents a genuine design flaw where similar APIs behave differently regarding error propagation.

After searching the repository, this appears to be the primary issue tracking this specific problem. Related error handling issues:

- #921 - Out of order resource error handling
- #4735 - No error handling for tedious error "ESOCKET" on connection  
- #1943 - Unusual behavior of promises when commit/rollback transaction errors are encountered

This issue should remain open and be prioritized for fixing in a future major version, as changing the behavior would be a breaking change. The fix should make `modify` consistent with other query builder methods by deferring callback execution.
